cpp/locality/NeighborComputeFunctional.h

```diff
@@ -0,0 +1,262 @@
+    \param qargs Query arguments
+    \param nlist Neighbor List. If not NULL, loop over it. Otherwise, use ref_points
+           appropriately with given qargs.
+    \param cf A void function that takes
+           (ref_point_index, point_index, distance, weight) as input.
+*/
+template<typename ComputePairType>
+void loopOverNeighbors(const NeighborQuery* ref_points, const vec3<float>* points, unsigned int Np,
+                       QueryArgs qargs, const NeighborList* nlist, const ComputePairType& cf,
+                       bool parallel = true)
+{
+    // check if nlist exists
+    if (nlist != NULL)
+    {
+        // if nlist exists, loop over it in parallel.
+        loopOverNeighborList(parallel, nlist, cf);
+    }
+    else
+    {
+        loopOverNeighborQuery(parallel, ref_points, points, Np, qargs, cf);
+    }
+}
+
+//! Wrapper iterating looping over NeighborList in parallel.
+/*! \param nlist Neighbor List to loop over.
+    \param cf A void function that takes
+           (ref_point_index, point_index, distance, weight) as input.
+*/
+template<typename ComputePairType>
+void loopOverNeighborList(bool parallel, const NeighborList* nlist, const ComputePairType& cf)
+{
+    const size_t* neighbor_list(nlist->getNeighbors());
+    size_t n_bonds = nlist->getNumBonds();
+    const float* neighbor_distances = nlist->getDistances();
+    const float* neighbor_weights = nlist->getWeights();
+    forLoopWrapper(parallel, 0, n_bonds, [=](size_t begin, size_t end) {
+        for (size_t bond = begin; bond != end; ++bond)
+        {
+            size_t i(neighbor_list[2 * bond]);
+            size_t j(neighbor_list[2 * bond + 1]);
+            cf(i, j, neighbor_distances[bond], neighbor_weights[bond]);
+        }
+    });
+}
+
+//! Wrapper iterating looping over NeighborQuery
+/*! \param ref_points NeighborQuery object to iterate over
+    \param points Points
+    \param Np Number of points
+    \param qargs Query arguments
+    \param cf A void function that takes
+           (ref_point_index, point_index, distance, weight) as input.
+*/
+template<typename ComputePairType>
+void loopOverNeighborQuery(bool parallel, const NeighborQuery* ref_points, const vec3<float>* points,
+                           unsigned int Np, QueryArgs qargs, const ComputePairType& cf)
+{
+    // if nlist does not exist, check if ref_points is an actual NeighborQuery
+    std::shared_ptr<NeighborQueryIterator> iter;
+    std::shared_ptr<AABBQuery> abq;
+    // check if ref_points is a pointer to a RawPoints object
+    // dynamic_cast will fail if ref_points is not actually pointing to RawPoints
+    // and return a null pointer. Then, the assignment operator will return
+    // a null pointer, making the condition in the if statement to be false.
+    // This is a typical C++ way of checking the type of a polymorphic class
+    // using pointers and casting.
+    if (const RawPoints* rp = dynamic_cast<const RawPoints*>(ref_points))
```

Comment:
The reason that I made them into separate functions is because of Cython typing issues, so I should figure out how this will work in a single function perhaps returning a tuple. 
Are you saying that we want to impose a restriction that if someone were to call `loopOverNeighborQuery`, the person should call with actual query object, not `RawPoints`?